pilot/pkg/model/push_context.go

```diff
@@ -860,9 +860,18 @@ func (ps *PushContext) GetAllServices() []*Service {
 }
 
 // ServiceForHostname returns the service associated with a given hostname following SidecarScope
+// It returns the service in the proxy namespace in high priority.
 func (ps *PushContext) ServiceForHostname(proxy *Proxy, hostname host.Name) *Service {
 	if proxy != nil && proxy.SidecarScope != nil {
-		return proxy.SidecarScope.servicesByHostname[hostname]
+		// first return the service from proxy namespace
+		if proxy.SidecarScope.servicesByHostname[hostname][proxy.ConfigNamespace] != nil {
+			return proxy.SidecarScope.servicesByHostname[hostname][proxy.ConfigNamespace]
+		}
+		// return arbitrary service
+		for _, svc := range proxy.SidecarScope.servicesByHostname[hostname] {
```

Comment:
Why does a sidecar need to store all the services? The intent behind SidecarScope is that it consists only of the resources _in scope_ for the proxy. These other hostnames are not in scope.

Reply:
after a second look, this seems not reasonable.

Reply:
@howardjohn updated comment, actually it is return a service of hostname from other namespaces, but it is also exported to this proxy.